### ISSUE TITLE
Cost increases during with training epochs 

### ISSUE THREAD

**Issue by flipdazed** [flipdazed/weather-modelling#2]
**Issue**
The cost should be decreasing with training epochs, instead it increases

**Affected**

    train_dbn.py
    train_mlp.py

**Reproduction**

    python train_mlp.py
    python train_dbn.py

**~~Runlog~~** _(open in OS X terminal with `cat runlog.txt` to enable full colour output)_

**Comment by flipdazed** [comment]
**Cost Calculation**

The cost is calculated by `models.rbm.RBM.getReconstructionCost` _(which is called by `models.rbm.RBM.getCostUpdates` as the argument `persistent=None`)_ given as,

    def getReconstructionCost(self, updates, pre_sigmoid_nv):
        cross_entropy = T.mean(
            T.sum(
                self.inputs * T.log(T.nnet.sigmoid(pre_sigmoid_nv)) +
                (1 - self.inputs) * T.log(1 - T.nnet.sigmoid(pre_sigmoid_nv)),
                axis=1
            )
        )
        return cross_entropy

with the `cross_entropy` described [here](http://deeplearning.net/tutorial/dA.html#daa) and mathematically expressed as,

\begin{equation}
L_{H}(\mathbb{x}, \mathbb{z}) 
= -\sum^d_{k=1}\left[
    \mathbb{x}_k \ln{\mathbb{z}_k} + 
    (1-\mathbb{x}_k)\ln{1-\mathbb{z}_k}
    \right]
\end{equation}

**Comment by flipdazed** [comment]
Checked for sign errors - see #3

**Comment by flipdazed** [comment]
The MNIST tests demonstrate **negative** costs **increasing** in magnitude i.e. the cost is decreasing as expected.

**Comment by flipdazed** [comment]
**nodes**
Here I tried `[15000, 15000, 15000]`. I assumed I would need `>10000` nodes in at least the first RBM layer to model cross-relationships between the data sets. I made this assumption from reading [this](http://www.cs.toronto.edu/~rsalakhu/papers/science.pdf) paper by Geoffrey Hinton who seemed to arbitrarily just bump up the first number of nodes depending on the complexity of the task at hand. I’m not entirely sure how complex this particular task is though and how many nodes to include and how to distribute them.

**learning rate, epochs, batch size, early stopping**
I don’t think this is the problem here as these would just cause the learning to be slower/faster rather than the wrong direction. Mini-batch size seems reasonable at 40 for a total sample size of 582. Early stopping is not part of the pre-training so can’t be responsible for the direction issues

**`k`, number of Gibbs steps in CD/PCD**
Is this possibly responsible? `1` is the lowest this could possibly be!

**Comment by flipdazed** [comment]
This seems to be a phenomenon distinctly of `models.rbm` - moved analysis of code to #1
